sdk/src/main/java/io/customer/sdk/di/DiGraph.kt

```diff
@@ -27,4 +27,35 @@ abstract class DiGraph {
      * Or you may not be able to successfully mock in test functions.
      */
     inline fun <reified DEP> override(): DEP? = overrides[DEP::class.java.simpleName] as? DEP
+
+    /**
+     * We prefer to have all of the SDK's singleton instances held in the dependency injection graph. This makes it easier for automated tests to be able to delete all
+     * singletons between each test function and prevent test flakiness.
+     */
+    val singletons: MutableMap<String, Any> = mutableMapOf()
```

Comment:
I changed the way that we create and store singles in the SDK. I had some failing automated tests and it was because singleton instances got created when the first unit test got executed and that same instance was used in all tests. I needed to delete all singleton instances after each test. 

So, to help automate this, I thought of this idea where all singletons are stored in the dependency graph so we can easily delete them all after each test. Fixed the issue I was having with tests not working.

Reply:
My only concern here would be what we have been discussing before, storing references in the `singletons` map. So even if that instance isn't being used it would still be referenced by `Digraph`?

Reply:
If in instance is saved in `Digraph.singletons` is not being used by another class then, yes, it would be referenced by `Digraph`. I currently can't think of any problems that could occur with `Digraph` containing all of these references. Can you think of any problems?

Reply:
I think it should be okay for now.